bin/strata-cli/src/cmd/faucet.rs

```diff
@@ -146,15 +95,60 @@ pub async fn faucet(args: FaucetArgs, seed: Seed, settings: Settings) {
             }
         };
 
-        println!("Claiming to signet address {}", address);
+        (address.to_string(), "claim_l1")
+    };
+
+    let client = reqwest::Client::new();
+    let base = Url::from_str(&settings.faucet_endpoint).expect("valid url");
+    let challenge = client
+        .get(base.join("/pow_challenge").unwrap())
+        .send()
+        .await
+        .unwrap()
+        .json::<PowChallenge>()
+        .await
+        .expect("invalid response");
+    println!(
+        "Received POW challenge with difficulty 2^{} from faucet: {:?}. Solving...",
+        challenge.difficulty, challenge.nonce
+    );
 
-        format!(
-            "{base}claim_l1/{}/{}",
-            encode(&solution.to_le_bytes()),
-            address
-        )
+    let mut solution = 0u64;
+    let prehash = {
+        let mut hasher = Sha256::new();
+        hasher.update(b"strata faucet 2024");
+        hasher.update(challenge.nonce.0);
+        hasher
     };
+    let pb = ProgressBar::new_spinner();
+    let mut counter = 0u64;
+    while !pow_valid(
+        prehash.clone(),
+        challenge.difficulty,
+        solution.to_le_bytes(),
+    ) {
+        solution += 1;
+        if counter % 100 == 0 {
+            pb.set_message(format!("Trying {solution}"));
+        }
+        counter += 1;
+    }
+    pb.finish_with_message(format!(
+        "✔ Solved challenge after {solution} attempts. Claiming now."
+    ));
 
+    #[cfg(feature = "strata_faucet")]
+    println!("Claiming to {} address {}", network_type, address);
+
+    #[cfg(not(feature = "strata_faucet"))]
+    println!("Claiming to signet address {}", address);
+
+    let url = format!(
+        "{base}{}/{}/{}",
+        claim,
+        encode(&solution.to_le_bytes()),
+        address
+    );
```

Comment:
A minor comment, but I'd say it make more sense to have this before the proof-of-work code (currently, line 100) as this refers to a previous context.